### PR TITLE
refactors hydroplasma guns

### PR DESCRIPTION
**Pull request by lolman360**
opening as draft because i have no idea if we have to modularize or we're only allowing straight ports rn

also it still needs testing

## About The Pull Request

- adds hydrogen cells to R&D, unlocked at the same time as other plasma weapons
- refactors plasma weapons completely. they now use datum/component/heat to tick heat, and as such are much better performance-wise (ostensibly). they also dont have FUCKED processing code anymore.
- plasma weapons take more skill to use, as they no longer automatically vent.

## Changelog
:cl:
add: hydrogen cells are individually printable now (previously, you'd need to print the gun to make one)
balance: hydrogen guns no longer instakill you on containment failure (lol 350 burn). no longer automatically vent either, so be careful! control-click vents them now.
balance: hydrogen guns do 20% less damage (100>80, they're still quite powerful)
spriteadd: plasma welder now looks distinct instead of just a purple-tanked basic welder
refactor: refactored plasmagun heat into a component, now we can add overheating anything
/:cl:

### CODE REVIEW

**Comment by r4d6** [comment]
As the original coder of the hydroplasma guns. I'll just say 2 things : 
1. You know you're supposed to buy the cells from cargo?
2. This make the hydrogen gun much easier to use and much more forgiving, when the concept was of a gun that's as unforgiving to you as to your enemies.

**Comment by AllCorruptingWyrm** [comment]
And as a commenter, I'll say two things;
1. You know we're changing things, right?
2. Unforgiving gameplay is all fun when you have a counterbalance and or a way to work around it.
Instant death mechanics aren't fun.

**Comment by lolman360** [comment]
While I appreciate your sentiment, your points don't really have any merit.

> 1. You know you're supposed to buy the cells from cargo?

Soteria can already print the cells in a somewhat roundabout way (printing the gun, extracting the cell from them), so why not printing the cells themselves? I can get ordering the backpack units, as AFAIK they can't be printed, and are much bigger. Additionally, the cells themselves require rare materials (mhydrogen, tritium, osmium) which are already unobtainable without Lonestar's smelter. So it's not like it makes them independent for the cells anyways.

> 2. This make the hydrogen gun much easier to use and much more forgiving, when the concept was of a gun that's as unforgiving to you as to your enemies.

I can understand this, and I might consider upping overheat damage on hands to ~30-40, depending on balance, but generally plasma guns are balanced even if heat is a total non-factor. Firstly, they require rare materials from Mining, are fairly slow-firing, cannot be reloaded in combat, and are quite expensive to reload even then. By adding overheat, to discourage continuous firing, and by adding the manual vent instead of the automatic 30-second vent, a user is both not as badly punished (instant kills are not fun) on mistakes, but can be rewarded with better heat management (i.e fire, fire, fire, wait, fire, vent, fire fire fire fire) instead of the gun just automatically venting at a certain point.

And finally: the original process proc on these. It is genuinely some of the jankiest code I've ever seen, sitting just below `prob(80) return`. For the sake of any codebase you help in the future, @r4d6, please please please use the processing subsystems (SSobj for 2 sec ticks, SSfastprocess for 1 sec ticks iirc) and never ever use `spawn()`- `addtimer` exists on every codebase for good  reason.

**Comment by r4d6** [comment]
> And finally: the original process proc on these. It is genuinely some of the jankiest code I've ever seen, sitting just below prob(80) return. For the sake of any codebase you help in the future, @r4d6, please please please use the processing subsystems (SSobj for 2 sec ticks, SSfastprocess for 1 sec ticks iirc) and never ever use spawn()- addtimer exists on every codebase for good reason.

That one is a fair point, back when I did the guns, I didn't know how the proc worked. I already put up a PR on sojourn to do it better.

**Comment by Cryleve** [comment]
Oh my a draft that isn't from Ziente. Let's see... I concur, I'd rather have the guns be easier and more rewarding to use. However, given the inspiration I'm more for a punishing explosion. Yet as it is this is preferable to instant obliteration if not more interactive and fun to use given your ability to manage heat manually. Leaning towards merging.

**Comment by lolman360** [comment]
actually that also sounds like a decent idea- a weak (light range 1-2) explosion on a short-ish delay, after the burn damage, so you have to throw it out of your hands or get exploded

**Comment by Cryleve** [comment]
As long as there is sufficient feedback that the weapon in your hands is about to discharge, go for it. Actually could be interesting to purposefully overheat your weapon to have it explode next to whatever you're fighting for knockback or etc.

**Comment by r4d6** [comment]
Well an explosion of Light Power 1 wouldn't actually hurt anyone other than the user. A minimum of 2 Power is needed for the explosion to affect other tile.
It is why the projectile make a Power 2 Explosion on impact, it is the smallest explosion that could still affect the target instead of just breaking the tile in front.

File: code/modules/projectiles/guns/plasma/hydrogen.dm [hunk 1]
```diff
@@ -38,27 +38,25 @@ Securing and unsecuring the flask is a long and hard task, and a failure when un
...
 	// Damage dealt when overheating
-	var/contain_fail_damage = 50 // Applied to every bodypart.
 	var/overheat_damage = 25 // Applied to the hand holding the gun.
 
-	var/aerith_aether = 50 // Variable used to repetidely call Process(), which is used for heat management. It is in deciseconds, so 50 = 5 seconds
 
 /obj/item/gun/hydrogen/Initialize()
 	..()
 	flask = new /obj/item/hydrogen_fuel_cell(src) // Give the gun a new flask when mapped in.
+	AddComponent(/datum/component/heat, clickTypeToVent = COMSIG_CLICK_CTRL, parentIsGun = TRUE, _heatThresholdSpecial = vent_level, _overheatThreshold = overheat, _heatPerFire = heat_per_shot, _coolPerTick = 2, _ventCooldown = vent_level_timer)
```
Comment (r4d6) [inline code/modules/projectiles/guns/plasma/hydrogen.dm:51]: Initialize(), unless it was explained to me in a very bad way, only called when the object is mapped in. With New() being the proc called upon creation.

Comment (lolman360) [reply code/modules/projectiles/guns/plasma/hydrogen.dm]: iirc, you're thinking about `Initialize(mapload)`- Init is called on `new`, and gives a controlled space to call procs that you'd normally call on `New()`, to deal with byond being a bit wierd.

Comment (r4d6) [reply code/modules/projectiles/guns/plasma/hydrogen.dm]: Well that certainly explain why you said that RnD could already print fuel cells in a round-about way.

The guns were supposed to be printed empty, and only be given a cell when mapped in.

**Comment by r4d6** [comment]
On a small unrelated note, I would love to see the new hydrogen welder sprite.

**Comment by r4d6** [comment]
I am currently porting this upstream @lolman360, but my only complain is that there doesn't seems to no longer be any consequences to firing the hydrogen gun while the cell is unsecured, so the long reload time is basically no longer there as this was the way to mechanically enforce this.

And considering I got no idea how the signal & component stuff work, I do not know how to add that part back in.

**Comment by lolman360** [comment]
right, i wasn't aware the gun could fire with an unsecured cell. 
dcs is a whole other kettle of fish, if you straight port everything after it's merged it should work

**Comment by r4d6** [comment]
Thanks a lot!

**Comment by lolman360** [comment]
image
what it looks like ingame atm

**Comment by r4d6** [comment]
Just tested it on my test server. Heat isn't going up and Ctrl+Click doesn't work.

**Comment by lolman360** [comment]
aware, issues are being resolved
heat isn't going up because the cooling values are too high, and control-click i have no idea (it should be sending a signal)

**Comment by r4d6** [comment]
> might fix plasma welders being weird on conversion

I... already fixed that part a bit ago : https://github.com/sojourn-13/sojourn-station/pull/2253

**Pull request merged**